admob/src/android/banner_view_internal_android.cc

```diff
@@ -192,96 +232,173 @@ void InitializeBannerViewOnMainThread(void* data) {
+  call_data->ad_unit_id = ad_unit_id;
+  call_data->ad_view = env->NewGlobalRef(ad_view_);
+  call_data->banner_view_helper = env->NewGlobalRef(helper_);
+  call_data->callback_data = callback_data;
+  util::RunOnMainThread(env, activity, InitializeBannerViewOnMainThread,
+                        call_data);
+
+  return MakeFuture(&future_data_.future_impl, callback_data->future_handle);
 }
 
-Future<LoadAdResult> BannerViewInternalAndroid::LoadAd(
-    const AdRequest& request) {
+// This function is run on the main thread and is called in the
+// BannerViewInternalAndroid::LoadAd() method.
+void LoadAdOnMainThread(void* data) {
+  LoadAdOnMainThreadData* call_data =
+      reinterpret_cast<LoadAdOnMainThreadData*>(data);
+  JNIEnv* env = GetJNI();
+  FIREBASE_ASSERT(env);
+
   admob::AdMobError error = kAdMobErrorNone;
-  jobject request_ref = GetJavaAdRequestFromCPPAdRequest(request, &error);
+  jobject j_ad_request =
+      GetJavaAdRequestFromCPPAdRequest(call_data->ad_request, &error);
 
-  if (request_ref == nullptr) {
+  if (j_ad_request == nullptr) {
     if (error == kAdMobErrorNone) {
       error = kAdMobErrorInternalError;
     }
-    return CreateAndCompleteFutureWithResult(
-        kBannerViewFnLoadAd, error, kAdCouldNotParseAdRequestErrorMessage,
-        &future_data_, LoadAdResult());
+    call_data->callback_data->future_data->future_impl.CompleteWithResult(
+        call_data->callback_data->future_handle, error,
+        kAdCouldNotParseAdRequestErrorMessage, LoadAdResult());
+    delete call_data->callback_data;
+    call_data->callback_data = nullptr;
+  } else {
+    ::firebase::admob::GetJNI()->CallVoidMethod(
+        call_data->banner_view_helper,
+        banner_view_helper::GetMethodId(banner_view_helper::kLoadAd),
+        reinterpret_cast<jlong>(call_data->callback_data), j_ad_request);
+    env->DeleteLocalRef(j_ad_request);
   }
+  delete call_data;
+}
+
+Future<LoadAdResult> BannerViewInternalAndroid::LoadAd(
+    const AdRequest& request) {
+  firebase::MutexLock lock(mutex_);
+
+  JNIEnv* env = GetJNI();
+  FIREBASE_ASSERT(env);
 
   FutureCallbackData<LoadAdResult>* callback_data =
       CreateLoadAdResultFutureCallbackData(kBannerViewFnLoadAd, &future_data_);
   SafeFutureHandle<LoadAdResult> future_handle = callback_data->future_handle;
 
-  ::firebase::admob::GetJNI()->CallVoidMethod(
-      helper_, banner_view_helper::GetMethodId(banner_view_helper::kLoadAd),
-      reinterpret_cast<jlong>(callback_data), request_ref);
+  LoadAdOnMainThreadData* call_data = new LoadAdOnMainThreadData();
+  call_data->ad_request = request;
+  call_data->callback_data = callback_data;
+  call_data->banner_view_helper = env->NewGlobalRef(helper_);
+
+  jobject activity = ::firebase::admob::GetActivity();
+  util::RunOnMainThread(env, activity, LoadAdOnMainThread, call_data);
 
   return MakeFuture(&future_data_.future_impl, future_handle);
 }
 
+BoundingBox BannerViewInternalAndroid::bounding_box() const {
```

Comment:
Method was renamed from BoundingBox BannerViewInternalAndroid::GetBoundingBox and moved up from below so that the methods appear in the same order as they're declared in the header file.